[WIP] Added Prototype for Exporting RV

ritesh99rakesh opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests . Please also
write a comment on that issue linking back to this pull request once it is
open. -->
Refer to the discussion: https://github.com/sympy/sympy/issues/16942


#### Brief description of what is fixed or changed
Added `_sample_numpy` and `_sample_scipy` for `Normal` distribution under
`sympy/stats/crv_types.py`

#### Other comments
Suggestions required.
Ping @Upabjojr @sidhantnagpal @czgdp1807 

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
NO ENTRY
<!-- END RELEASE NOTES -->

ritesh99rakesh committed:
Added Prototype for Exporting RV

ritesh99rakesh commented:
@Upabjojr I have added the sample function using numpy and scipy for normal distribution. But as you can see, there is a lot of code duplication.
Therefore, I feel that we should discuss the exact nature of the functions and place where we should keep them.
Please review.

czgdp1807 commented on sympy/stats/crv_types.py:
```diff
@@ -2320,6 +2320,32 @@ def pdf(self, x):
 
+    def _sample_numpy(self, **kwargs):
+        mean, std = self.mean, self.std
+        if isinstance(mean, Symbol) or isinstance(std, Symbol):
```
Just a suggestion for the future work, use, something like, `mean.has(Symbol) or std.has(Symbol)`. No need to make changes now, as this is a prototype.

czgdp1807 commented on sympy/stats/crv_types.py:
```diff
@@ -2320,6 +2320,32 @@ def pdf(self, x):
+
+        numpy = import_module('numpy')
+        if numpy:
+            mean, std = float(mean), float(std)
```
Casting may invite errors in computation. Is there any other method for going from `SymPy` objects to `float` objects.

czgdp1807 replied:
Like, using `args` and all that stuff. Please try to remove casting, if possible.

ritesh99rakesh replied:
I don't have much idea in this area. Maybe @sidhantnagpal clarify this.

czgdp1807 replied:
May be there isn't a way. I tried it out. Comment if you can find out any, otherwise, it's fine to do casting. I think, `SymPy` has its own method like, `__float__` which can avoid errors. Can you check?

ritesh99rakesh replied:
`__float__` method internally uses `float` to convert to float.
```python
def __float__(self):
    # Don't bother testing if it's a number; if it's not this is going
    # to fail, and if it is we still need to check that it evalf'ed to
    # a number.
    result = self.evalf()
    if result.is_Number:
        return float(result)
    if result.is_number and result.as_real_imag()[1]:
        raise TypeError("can't convert complex to float")
    raise TypeError("can't convert expression to float")
```
So, I think above is fine.

czgdp1807 replied:
Resolve this conversation.

sidhantnagpal commented:
Much of the duplication can be removed if you factor out the common logic (like ensuring numeric `mu`, `sigma` and `import_module` checks) and move it to `sample` method (which could itself be moved to superclass).

This is assuming `_sample_*` methods are planned to be accessed only through `sample`.

ritesh99rakesh commented:
> This is assuming `_sample_*` methods are planned to be accessed only through `sample`.

Yes @sidhantnagpal I believe that `_sample_*` methods are supposed to called internally only through `sample`. @Upabjojr what do you think?

ritesh99rakesh committed:
Reduced code duplication

ritesh99rakesh commented:
@sidhantnagpal what do you think about this approach?

sidhantnagpal commented on sympy/stats/crv_types.py:
```diff
@@ -2317,8 +2317,33 @@ def check(mean, std):
-        return random.normalvariate(self.mean, self.std)
+    def sample(self, **kwargs):
+        mean, std = self.mean, self.std
+        print("len: ", len(kwargs))
```
This should be removed.

sidhantnagpal commented on sympy/stats/crv_types.py:
```diff
@@ -2317,8 +2317,33 @@ def check(mean, std):
+        objs = {}
+        mean, std, size = float(mean), float(std), kwargs.get('size', 1)
+        if kwargs.get('numpy', False):
+            objs['numpy'] = _sample_numpy(mean, std, size)
```
Don't you intend to return these?

ritesh99rakesh replied:
Yes. I will make the correction.

Upabjojr replied:
If you are calling a class method, you should prepend `self.` to the method name.

Upabjojr commented on sympy/stats/crv_types.py:
```diff
@@ -2317,8 +2317,33 @@ def check(mean, std):
+    def _sample_numpy(mean, std, size):
+        numpy = import_module('numpy')
+        if numpy:
+            return numpy.random.normal(mean, std, size)
```
this method should just return, no module import.

Upabjojr commented on sympy/stats/crv_types.py:
```diff
@@ -2317,8 +2317,33 @@ def check(mean, std):
 
-    def sample(self):
-        return random.normalvariate(self.mean, self.std)
+    def sample(self, **kwargs):
```
Why is this here? It should belong to the superclass.

Upabjojr commented:
> I believe that `_sample_*` methods are supposed to called internally only through `sample`. @Upabjojr what do you think?

Yes, that's the purpose of those methods.

Upabjojr commented:
Where are the tests? I had a glance at your code and it really looks like it cannot work in the current state.

Upabjojr commented:
@ritesh99rakesh please make this PR work. I would recommend transforming the currently existing `sample` methods into `_sample_python`, then define `sample` in the superclass so that it (temporarily) calls `_sample_python`, with the existing code.

As soon as the other methods are ready we can start adding a dispatcher for the other methods.

Please refrain temporarily from writing a dispatcher (e.g. to choose the methods). We will do it later, when the sampler are working.

Can you do this by tomorrow?

ritesh99rakesh commented:
> Can you do this by tomorrow?

Yes, I will be able to complete it by today and will ask for your suggestions on the code.

ritesh99rakesh committed:
Merge branch 'master' into sample

ritesh99rakesh committed:
Added _sample_python, made the code working

ritesh99rakesh commented:
@Upabjojr Following changes were made:
1. Added `_sample_python` to `NormalDistribution`.
2. Removed `_inverse_cdf_expression` function since we do not require it to find the sample.

As you can see that `_sample_python` in itself is not much dependent upon `NormalDistribution` and could be made generic by shifting the code to `sample` under `stats/crv.py`.

I look forward to your suggestions.

Upabjojr commented:
>     2\. Removed `_inverse_cdf_expression` function since we do not require it to find the sample.

Why are you removing working code?!?!?!?!?

I just asked you to move the current `sample` function into `_sample_python`. Do not remove anything.

Upabjojr commented:
We are going to add extra sampling functions (like `_sample_scipy`, `_sample_pymc3`, `_sample_stan`, etc...)

ritesh99rakesh commented:
> I just asked you to move the current `sample` function into `_sample_python`. Do not remove anything.

Ok @Upabjojr! I will keep that in mind and will restore this.

ritesh99rakesh committed:
Added pymc3 and refactored code to make more generic

ritesh99rakesh committed:
Merge origin

sidhantnagpal commented on sympy/stats/crv.py:
```diff
@@ -169,15 +169,15 @@ def __new__(cls, *args):
         """ A random realization from the distribution """
+        return self._sample_python(**kwargs)
         icdf = self._inverse_cdf_expression()
         return icdf(random.uniform(0, 1))
```
This code is not reachable due to early return.

ritesh99rakesh replied:
Yes, I know. I have added my code without deleting the previous code so that we can finalise the prototype.
Also, could you tell me your review about returning the `objs` in `NormalDistribution` under `sympy/stats/crv_types.py`? The current return type is `object`. Is there a better way?

Upabjojr commented on sympy/stats/crv_types.py:
```diff
@@ -2339,9 +2339,60 @@ def check(mean, std):
         return exp(-(x - self.mean)**2 / (2*self.std**2)) / (sqrt(2*pi)*self.std)
 
-    def sample(self):
+    def _sample_python(self, **kwargs):
```
This method is supposed to be the only `sample` method, just that.

Upabjojr commented on sympy/stats/rv.py:
```diff
@@ -369,7 +369,7 @@ def domain(self):
         raise NotImplementedError("Density not available for ProductSpaces")
 
-    def sample(self):
+    def sample(self, **kwargs):
```
why `**kwargs`?

Upabjojr commented:
> Where are the tests? I had a glance at your code and it really looks like it cannot work in the current state.

I still see no tests. Where do you test the new methods you have added?

Upabjojr commented:
I'm closing this PR as it's becoming quite a mess.

Please start over. Let's have two PRs:

- addition of new methods (e.g. `_sample_pymc3`, `_sample_scipy`). Just that, no dispatchers for now.
- moving `sample` to `_samply_python`, just that.

Upabjojr commented:
And espacially, please add some tests to every PR.
